pkg/nodeagent/controller/operatingsystemconfig/containerd.go

```diff
@@ -0,0 +1,363 @@
+	if err != nil {
+		return fmt.Errorf("unable to write unit dropin: %w", err)
+	}
+
+	return nil
+}
+
+// ensureContainerdConfiguration sets the configuration for containerd.
+func (r *Reconciler) ensureContainerdConfiguration(criConfig *extensionsv1alpha1.CRIConfig) error {
+	config, err := r.FS.ReadFile(configFile)
+	if err != nil {
+		return fmt.Errorf("unable to read containerd config.toml: %w", err)
+	}
+
+	content := map[string]any{}
+
+	if err = toml.Unmarshal(config, &content); err != nil {
+		return fmt.Errorf("unable to decode containerd default config: %w", err)
+	}
+
+	type (
+		patch struct {
+			name  string
+			path  structuredmap.Path
+			setFn structuredmap.SetFn
+		}
+	)
+
+	patches := []patch{
+		{
+			name: "registry config path",
+			path: structuredmap.Path{"plugins", "io.containerd.grpc.v1.cri", "registry", "config_path"},
+			setFn: func(_ any) (any, error) {
+				return certsDir, nil
+			},
+		},
+		{
+			name: "imports paths",
+			path: structuredmap.Path{"imports"},
+			setFn: func(value any) (any, error) {
+				importPath := path.Join(configDir, "*.toml")
+
+				imports, ok := value.([]any)
+				if !ok {
+					return []string{importPath}, nil
+				}
+
+				for _, imp := range imports {
+					path, ok := imp.(string)
+					if !ok {
+						continue
+					}
+
+					if path == importPath {
+						return value, nil
+					}
+				}
+
+				return append(imports, importPath), nil
+			},
+		},
+		{
+			name: "sandbox image",
+			path: structuredmap.Path{"plugins", "io.containerd.grpc.v1.cri", "sandbox_image"},
+			setFn: func(value any) (any, error) {
+				if criConfig == nil || criConfig.Containerd == nil {
+					return value, nil
+				}
+
+				return criConfig.Containerd.SandboxImage, nil
+			},
+		},
+	}
+
+	for _, p := range patches {
+		content, err = structuredmap.SetMapEntry(content, p.path, p.setFn)
+		if err != nil {
+			return fmt.Errorf("unable setting %q in containerd config.toml: %w", p.name, err)
+		}
+	}
+
+	f, err := r.FS.OpenFile(configFile, os.O_CREATE|os.O_WRONLY|os.O_TRUNC, 0644)
+	if err != nil {
+		return fmt.Errorf("unable to open containerd config.toml: %w", err)
+	}
+	defer func() {
+		err = f.Close()
+	}()
+
+	return toml.NewEncoder(f).Encode(content)
+}
+
+// ensureContainerdRegistries configures containerd to use the desired image registries.
+func (r *Reconciler) ensureContainerdRegistries(ctx context.Context, log logr.Logger, newRegistries []extensionsv1alpha1.RegistryConfig) <-chan error {
+	var (
+		errChan = make(chan error, 1)
+
+		registriesWithoutReadiness []extensionsv1alpha1.RegistryConfig
+		registriesWithReadiness    []extensionsv1alpha1.RegistryConfig
+	)
+
+	for _, registryConfig := range newRegistries {
+		if ptr.Deref(registryConfig.ReadinessProbe, false) {
+			registriesWithReadiness = append(registriesWithReadiness, registryConfig)
+		}
+		registriesWithoutReadiness = append(registriesWithoutReadiness, registryConfig)
+	}
```

Comment:
If registries with `ReadinessProbe=true` are added to `registriesWithoutReadiness`, this may cause `ReconcileContainerdRegistries` to fail 
```suggestion
	for _, registryConfig := range newRegistries {
		if ptr.Deref(registryConfig.ReadinessProbe, false) {
			registriesWithReadiness = append(registriesWithReadiness, registryConfig)
		} else {
		  registriesWithoutReadiness = append(registriesWithoutReadiness, registryConfig)
		}
	}
```

Reply:
🚀

Reply:
🙈 thanks! 😄